packages/react-utils/src/useDismiss.ts

```diff
@@ -0,0 +1,112 @@
+'use client'
+
+import type { HintedString } from '@noaignite/types'
+import { useCallback, useEffect, useRef, type RefObject } from 'react'
+import { useEvent } from './useEvent'
+import { useIsomorphicEffect } from './useIsomorphicEffect'
+
+export type DismissOptions = {
+  /**
+   * A `boolean` indicating whether to listen for `keydown` events involving `Escape` key.
+   *
+   * @defaultValue true
+   */
+  keyboard?: boolean
+  /**
+   * A `boolean` indicating whether to listen for `pointerdown` events.
+   * Can be set to listen for any pointer type, or restricted to a specific type(s).
+   *
+   * @defaultValue true
+   */
+  pointer?:
+    | boolean
+    | HintedString<'mouse' | 'touch' | 'pen'>
+    | HintedString<'mouse' | 'touch' | 'pen'>[]
+  /**
+   * A `boolean` indicating whether the hook is enabled.
+   *
+   * @defaultValue true
+   */
+  when?: boolean
+}
+
+/**
+ * Executes `callback` when a `keydown` event involving the `Escape` key or a `pointerdown` event
+ * occurs outside of passed `ref` element. Useful when creating components, like dialogs, popovers,
+ * dropdowns, etc., that should be dismissed when the user interacts outside of them.
+ *
+ * @param ref - Element reference to base events on.
+ * @param callback - Callback to execute when a `keydown` or `pointerdown` event occurs.
+ * @param options - Configurable options
+ *
+ * @example
+ * ```tsx
+ * const ref = useRef<HTMLDivElement>(null)
+ * const [isOpen, setIsOpen] = useState(false)
+ *
+ * useDismiss(ref, setIsOpen.bind(null, false), { when: isOpen })
```

Comment:
todo: I agree with you that `bind` can make it cleaner in the implementation. I do think though that it would be more clear to not use `.bind()` in the example :)